src/ephy-sync-service.c

```diff
@@ -0,0 +1,1417 @@
+          ephy_bookmarks_manager_add_tag (manager, g_sequence_get (iter));
+      } else {
+        if (ephy_bookmark_get_modified (local) > ephy_bookmark_get_modified (remote))
+          ephy_sync_service_upload_bookmark (service, local, TRUE);
+
+        g_object_unref (remote);
+      }
+    }
+  }
+
+handle_local_bookmarks:
+  for (iter = g_sequence_get_begin_iter (bookmarks);
+       !g_sequence_iter_is_end (iter); iter = g_sequence_iter_next (iter)) {
+    EphyBookmark *bookmark = EPHY_BOOKMARK (g_sequence_get (iter));
+
+    if (ephy_bookmark_is_uploaded (bookmark) == TRUE)
+      ephy_sync_service_delete_bookmark (service, bookmark, TRUE);
+    else
+      ephy_sync_service_upload_bookmark (service, bookmark, FALSE);
+  }
+
+  /* Save changes to file. */
+  ephy_bookmarks_manager_save_to_file_async (manager, NULL, NULL, NULL);
+
+  /* Set the sync time. */
+  timestamp = soup_message_headers_get_one (msg->response_headers, "X-Weave-Timestamp");
+  server_time = g_ascii_strtod (timestamp, NULL);
+  ephy_sync_service_set_sync_time (service, server_time);
+
+out:
+  g_object_unref (parser);
+
+  ephy_sync_service_release_next_storage_message (service);
+}
+
+void
+ephy_sync_service_sync_bookmarks (EphySyncService *self,
+                                  gboolean         first)
+{
+  char *endpoint;
+
+  g_return_if_fail (EPHY_IS_SYNC_SERVICE (self));
+  g_return_if_fail (ephy_sync_service_is_signed_in (self));
+
+  endpoint = g_strdup_printf ("storage/%s?full=true", EPHY_BOOKMARKS_COLLECTION);
+
+  if (first == TRUE) {
+    ephy_sync_service_send_storage_message (self, endpoint,
+                                            SOUP_METHOD_GET, NULL, -1, -1,
+                                            sync_bookmarks_first_time_response_cb, NULL);
+  } else {
+    ephy_sync_service_send_storage_message (self, endpoint,
+                                            SOUP_METHOD_GET, NULL,
+                                            ephy_sync_service_get_sync_time (self), -1,
+                                            sync_bookmarks_response_cb, NULL);
+  }
+
+  g_free (endpoint);
+}
+
+static gboolean
+do_periodical_sync (gpointer user_data)
+{
+  EphySyncService *service = EPHY_SYNC_SERVICE (user_data);
+
+  ephy_sync_service_sync_bookmarks (service, FALSE);
+
+  return G_SOURCE_CONTINUE;
+}
+
+void
+ephy_sync_service_start_periodical_sync (EphySyncService *self,
+                                         gboolean         now)
+{
+  g_return_if_fail (EPHY_IS_SYNC_SERVICE (self));
+
+  if (ephy_sync_service_is_signed_in (self) == FALSE)
+    return;
+
+  if (now == TRUE)
+    do_periodical_sync (self);
+
+  self->source_id = g_timeout_add_seconds (SYNC_FREQUENCY, do_periodical_sync, self);
+}
+
+void
+ephy_sync_service_stop_periodical_sync (EphySyncService *self)
+{
+  g_return_if_fail (EPHY_IS_SYNC_SERVICE (self));
+
+  if (ephy_sync_service_is_signed_in (self) == FALSE)
+    return;
+
+  g_source_remove (self->source_id);
```

Comment:
Things go wrong here if the function gets called twice in a row (you either attempt to remove a nonexistent source, or the source ID has been recycled already and you remove some poor unrelated source!), which you would expect to be safe. Be careful to check if the source is nonzero before removing it, and zero it out once that's done.